### ISSUE TITLE
Batched greedy TDT records a stale token_duration, so end timestamps depend on batch composition

### ISSUE THREAD

**Issue by udsy19** [NVIDIA-NeMo/NeMo#16097]
**Describe the bug**

For TDT models, `GreedyBatchedTDTInfer` reports a `token_duration` that depends on which *other*
utterances happen to be in the same batch. Because `_compute_offsets_tdt` computes
`end_offset = start_offset + token_duration`, the token / word / segment **end timestamps** of an
utterance change with batch composition, even though the transcript and the start offsets do not.

This affects `GreedyBatchedTDTLabelLoopingComputer.torch_impl` — the path used on CPU, when
`use_cuda_graph_decoder=False`, when CUDA conditional-node support is missing, and while CUDA graphs
are disabled for validation decoding (dispatch at `label_looping_base.py:320-335`). The CUDA graph
implementation is not affected.

Root cause, `nemo/collections/asr/parts/submodules/transducer_decoding/tdt_label_looping.py`,
stage 1.2 inner loop of `torch_impl`:

```python
# labels and scores are written back only for elements still searching:
torch.where(advance_mask, more_labels, labels, out=labels)     # line 532
torch.where(advance_mask, more_scores, scores, out=scores)     # line 534

# durations are rebound for the whole batch:
jump_durations_indices = logits[:, -num_durations:].argmax(dim=-1)   # line 535
durations = model_durations[jump_durations_indices]                  # line 536
```

Stage 2 then stores `token_durations=durations` (line 575) for every utterance that found a label —
including ones that found it in an earlier iteration, whose duration has since been overwritten by a
joint evaluation at a later frame. The inner loop only iterates while some *other* utterance is still
searching, hence the batch dependence.

**The CUDA graph implementation of the same algorithm already has the guard**, at line 1352 of the
same file, in `_inner_loop_step_find_next_non_blank`:

```python
torch.where(self.state.advance_mask, more_durations, self.state.durations, out=self.state.durations)
```

So the two implementations of one algorithm simply disagree here.

`rnnt_decoding.py:341` sets `tdt_include_token_duration = tdt_include_token_duration or compute_timestamps`,
so simply asking for timestamps enables this field, and `rnnt_decoding.py:1203-1207` turns it into
`end_offset`.

**Steps/Code to reproduce bug**

CPU-only, no downloads.

```python
import torch
import nemo.collections.asr.parts.submodules.rnnt_greedy_decoding as gd
from nemo.collections.asr.modules import RNNTDecoder, RNNTJoint

TOKENS = [" ", "a", "b", "c"]
V, BLANK, D = len(TOKENS), len(TOKENS), 4
DUR = [0, 1, 2, 4]

torch.manual_seed(0)
dec = RNNTDecoder({'pred_hidden': D, 'pred_rnn_layers': 1}, V)
jnt = RNNTJoint({'encoder_hidden': D, 'pred_hidden': D, 'joint_hidden': D, 'activation': 'relu'},
                V, vocabulary=TOKENS, num_extra_outputs=len(DUR))
dec.eval(); jnt.eval()

g = torch.Generator().manual_seed(136)
enc = torch.randn(4, D, 12, generator=g)
lens = torch.randint(low=1, high=13, size=(4,), generator=g, dtype=torch.int32)
lens[0] = 12  # -> [12, 4, 9, 5]

def decode(e, l):
    d = gd.GreedyBatchedTDTInfer(dec, jnt, blank_index=BLANK, durations=DUR,
                                 max_symbols_per_step=10, include_duration=True,
                                 use_cuda_graph_decoder=False)
    with torch.no_grad():
        return [(list(map(int, h.y_sequence)), list(map(int, h.timestamp)),
                 list(map(int, h.token_duration)))
                for h in d(encoder_output=e, encoded_lengths=l)[0]]

full = decode(enc, lens)
for i in range(4):
    alone = decode(enc[i:i+1], lens[i:i+1])[0]
    print(i, "alone", alone, "| in batch", full[i])
```

Output — tokens and start timestamps match, `token_duration` does not:

```
0 alone ([1, 1], [0, 7], [1, 1]) | in batch ([1, 1], [0, 7], [4, 1])
1 alone ([1],    [0],    [1])    | in batch ([1],    [0],    [4])
2 alone ([1, 1], [0, 1], [1, 1]) | in batch ([1, 1], [0, 1], [1, 4])
3 alone ([],     [],     [])     | in batch ([],     [],     [])
```

Through `_compute_offsets_tdt` those become `end_offset` values of `[1, 8] / [1] / [1, 2]` when
decoded alone and `[4, 8] / [4] / [1, 5]` in a batch.

The batch-of-1 values are the correct ones. Evaluating the joint directly at frame 0 with the SOS
state for utterance 1 gives

```
token logits    = [-3.1643, -1.5967, -3.0091, -2.7906, -1.9962]  -> token 1
duration logits = [-2.2482, -1.7770, -2.2816, -2.0268]  -> index 1 -> duration 1
```

i.e. duration 1, which is what the batch-of-1 decode reports and what the batch-of-4 decode does not.

**Expected behavior**

`token_duration` (and therefore the TDT end timestamps) should be the duration predicted together
with the emitted token, and must not depend on the other utterances in the batch.

Applying the same guard the CUDA graph implementation uses fixes it:

```python
jump_durations_indices = logits[:, -num_durations:].argmax(dim=-1)
more_durations = model_durations[jump_durations_indices]
torch.where(advance_mask, more_durations, durations, out=durations)
```

With that change, all four utterances report the same `token_duration` alone and in a batch.

**Environment overview**

- Environment location: Bare-metal (macOS arm64), CPU only
- Method of NeMo install: from source, `main` @ `d0a07c056`

**Environment details**

- PyTorch 2.12.0
- Python 3.10.18

**Additional context**

Possibly related but distinct: PR #14912 fixes TDT **beam search** timestamps (`BatchedBeamHyps` not
populating `token_duration` at all, plus START/END timestamp semantics in `_compute_offsets_tdt`).
It touches `rnnt_decoding.py` and `batched_beam_decoding_utils.py`, not `tdt_label_looping.py`. This
report is about the greedy batched `torch_impl`, which populates the field but populates it with a
stale value.

Separately, in the `record_all_steps` branch the two implementations pass different values to
`add_results_masked_`: `torch_impl` line 544 passes the newly computed durations, while the CUDA graph
`_inner_loop_step_find_next_non_blank` line 1328 passes `self.state.durations` from before the
line-1352 update. Those two probably should agree as well, but that path requires
`preserve_alignments` / non-excluded blank confidence and is not what this report is about.
